gitops/gitops.go

```diff
@@ -0,0 +1,97 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitops
+
+import (
+	"fmt"
+	"os/exec"
+	"path/filepath"
+
+	appstudiov1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
+	"github.com/spf13/afero"
+)
+
+type Executor interface {
+	Execute(baseDir, command string, args ...string) ([]byte, error)
+}
+
+// GenerateAndPush takes in the flollowing args and generates the gitops resources for a given component
+// 1. outputPath: Where
+// 2. remote: A string of the form https://$token@github.com/<org>/<repo>. Corresponds to the component's gitops repository
+// 2. component: A component struct corresponding to a single Component in an Application in AS
+// 4. The executor to use to execute the git commands (either gitops.executor or gitops.mockExecutor)
+// 5. The filesystem object used to create (either ioutils.NewFilesystem() or ioutils.NewMemoryFilesystem())
+// 6. The branch to push to
+// 7. The path within the repository to generate the resources in
+// Adapted from https://github.com/redhat-developer/kam/blob/master/pkg/pipelines/utils.go#L79
+func GenerateAndPush(outputPath string, remote string, component appstudiov1alpha1.Component, e Executor, appFs afero.Fs, branch string, context string) error {
+	componentName := component.Name
+	if out, err := e.Execute(outputPath, "git", "clone", remote, componentName); err != nil {
+		return fmt.Errorf("failed to clone git repository in %q %q: %s", outputPath, string(out), err)
+	}
+
+	repoPath := filepath.Join(outputPath, componentName)
+
+	// Checkout the specified branch
+	if _, err := e.Execute(repoPath, "git", "switch", branch); err != nil {
+		if out, err := e.Execute(repoPath, "git", "checkout", "-b", branch); err != nil {
+			return fmt.Errorf("failed to checkout branch %q in %q %q: %s", branch, repoPath, string(out), err)
+		}
+	}
+
+	if out, err := e.Execute(repoPath, "rm", "-rf", filepath.Join("components", componentName)); err != nil {
+		return fmt.Errorf("failed to delete %q folder in repository in %q %q: %s", filepath.Join("components", componentName), repoPath, string(out), err)
+	}
+
+	// Generate the gitops resources
+	componentPath := filepath.Join(repoPath, context, "components", componentName, "base")
+	generate(appFs, componentPath, component)
+
+	if out, err := e.Execute(repoPath, "git", "add", "."); err != nil {
+		return fmt.Errorf("failed to add pipelines.yaml to repository in %q %q: %s", repoPath, string(out), err)
+	}
+
+	// See if any files changed, and if so, commit and push them up to the repository
+	if out, err := e.Execute(repoPath, "git", "--no-pager", "diff", "--cached"); err != nil {
+		return fmt.Errorf("failed to check git diff in repository %q %q: %s", repoPath, string(out), err)
+	} else if string(out) != "" {
```

Comment:
wonder if it should be just this 
```suggestion
	} else  {
```

Reply:
Not quite. If we do the generation and no files changed, there's point in doing an empty commit and pushing up to git. (It also means we have to add a `--allow-empty` flag to `git commit`)